src/python_testing/mdns_discovery/mdns_async_service_info.py

```diff
@@ -0,0 +1,273 @@
+                        # wait at least the duplicate question interval
+                        # before asking another QM question otherwise
+                        # its likely to be suppressed by the question
+                        # history of the remote responder.
+                        delay = _DUPLICATE_QUESTION_INTERVAL
+
+                await self.async_wait(min(next_, last) - now, self._zc.loop)
+                now = current_time_millis()
+        finally:
+            self._zc.async_remove_listener(self)
+
+        return True
+
+    def async_add_listener(
+        self, listener: RecordUpdateListener, question: Optional[Union[DNSQuestion, List[DNSQuestion]]]
+    ) -> None:
+        """Adds a listener for a given question.  The listener will have
+        its update_record method called when information is available to
+        answer the question(s).
+
+        This function is not threadsafe and must be called in the eventloop.
+        """
+        self._zc.record_manager.async_add_listener(listener, question)
+
+    def _generate_request_query(
+        self, zc: 'Zeroconf', now: float_, question_type: DNSQuestionType, record_type: DNSRecordType
+    ) -> DNSOutgoing:
+        """Generate the request query."""
+        out = DNSOutgoing(_FLAGS_QR_QUERY)
+        name = self._name
+        server = self.server or name
+        qu_question = question_type is QU_QUESTION
+        if record_type is None or record_type is DNSRecordType.SRV:
+            print("Requesting MDNS SRV record...")
+            self._add_question(
+                out, qu_question, server, _TYPE_SRV, _CLASS_IN
+            )
+        if record_type is None or record_type is DNSRecordType.TXT:
+            print("Requesting MDNS TXT record...")
+            self._add_question(
+                out, qu_question, server, _TYPE_TXT, _CLASS_IN
+            )
+        if record_type is None or record_type is DNSRecordType.A:
+            print("Requesting MDNS A record...")
+            self._add_question(
+                out, qu_question, server, _TYPE_A, _CLASS_IN
+            )
+        if record_type is None or record_type is DNSRecordType.AAAA:
+            print("Requesting MDNS AAAA record...")
+            self._add_question(
+                out, qu_question, server, _TYPE_AAAA, _CLASS_IN
+            )
+        return out
+
+    def _add_question(
+        self,
+        out: DNSOutgoing,
+        qu_question: bool,
+        name: str_,
+        type_: int_,
+        class_: int_
+    ) -> None:
+        """Add a question."""
+        question = DNSQuestion(name, type_, class_)
+        question.unicast = qu_question
+        out.add_question(question)
+
+    def _get_initial_delay(self) -> float_:
+        return _LISTENER_TIME
+
+    def _get_random_delay(self) -> int_:
+        return randint(*_AVOID_SYNC_DELAY_RANDOM_INTERVAL)
+
+    def _set_properties(self, properties: Dict[Union[str, bytes], Optional[Union[str, bytes]]]) -> None:
```

Comment:
why is this code an option? This class is owned by the caller, right? So why use a string dict to set properties when the properties could just be set?

Reply:
This code is part of a method from a zeroconf class that has been overridden, though this particular code isn't the part being personalized, this executes as is in the zeroconf library overridden (in this case) or not.